lib/Imagine/Image/AbstractImage.php

```diff
@@ -63,33 +68,60 @@ public function thumbnail(BoxInterface $size, $mode = ImageInterface::THUMBNAIL_
             $ratio = max($ratios);
         }
 
-        if ($mode === ImageInterface::THUMBNAIL_OUTBOUND) {
+        if ($mode === ImageInterface::THUMBNAIL_INSET) {
+            $imageSize = $imageSize->scale($ratio);
+            $thumbnail->resize($imageSize, $filter);
+        } else {
             if (!$imageSize->contains($size)) {
+                if ($allowUpscale) {
+                    $imageSize = $imageSize->scale($ratio);
+                    $thumbnail->resize($imageSize, $filter);
+                }
                 $size = new Box(
                     min($imageSize->getWidth(), $size->getWidth()),
                     min($imageSize->getHeight(), $size->getHeight())
                 );
             } else {
-                $imageSize = $thumbnail->getSize()->scale($ratio);
+                $imageSize = $imageSize->scale($ratio);
                 $thumbnail->resize($imageSize, $filter);
             }
             $thumbnail->crop(new Point(
                 max(0, round(($imageSize->getWidth() - $size->getWidth()) / 2)),
                 max(0, round(($imageSize->getHeight() - $size->getHeight()) / 2))
             ), $size);
-        } else {
-            if (!$imageSize->contains($size)) {
-                $imageSize = $imageSize->scale($ratio);
-                $thumbnail->resize($imageSize, $filter);
-            } else {
-                $imageSize = $thumbnail->getSize()->scale($ratio);
-                $thumbnail->resize($imageSize, $filter);
-            }
         }
 
         return $thumbnail;
     }
 
+    /**
+     * Check the settings argument in thumbnail() method.
+     *
+     * @param int $settings
+     */
+    private function checkThumbnailSettings($settings)
+    {
+        // Preserve BC until version 1.0
+        if ($settings === 'inset') {
+            $settings = ImageInterface::THUMBNAIL_INSET;
+        } elseif ($settings === 'outbound') {
+            $settings = ImageInterface::THUMBNAIL_OUTBOUND;
+        }
+
+        $allSettings = ImageInterface::THUMBNAIL_INSET | ImageInterface::THUMBNAIL_OUTBOUND | ImageInterface::THUMBNAIL_UPSCALE;
+
+        if (!is_int($settings) || ($settings & ~$allSettings)) {
+            throw new InvalidArgumentException('Invalid setting specified');
+        }
+
+        if ($settings & ImageInterface::THUMBNAIL_INSET &&
+            $settings & ImageInterface::THUMBNAIL_OUTBOUND) {
+            throw new InvalidArgumentException('Only one mode should be specified');
+        }
+
+        return $settings;
```

Comment:
@vlakoff Shouldn't $settings contain one of THUMBNAIL_INSET or THUMBNAIL_OUTBOUND?

Reply:
In `thumbnails()`, the `$settings` parameter has a default value, and more relevant, it's actually supported to omit the mode, see my next comment.

Reply:
I'm already handling all the probable errors, and the code is already quite complex. So I tried to not complexify the code more than needed.

This being said, `$settings` being a bitmask, a value of `0` is not that wrong, it just means "specify no flag". Still, I'd tend to agree this value is not supposed to be used. But I think adding code to reject it is not worth it.

That's because, how I thought the API, this 2nd parameter `$settings` is supposed to be "omittable" (calling the method with only the 1st parameter), but not "skippable" (if you use the 3rd parameter, you have to provide a meaningful value for this 2nd parameter). So I consider values like `0` or `null` to be unexpected and unsupported.

Reply:
A solution could be to add a `THUMBNAIL_DEFAULT = 0` constant, that could be used to skip the parameter. But I think the added complexity is not worth it. I think it's better to require the user to specify a meaningful value for the 2nd parameter if he uses the 3rd parameter, and consider `0` as an unsupported value (i.e. works, but unexpected).

This being said, you are free to tweak the PR if you wish, I'm fine with it.